Segfault in tether

https://ci.vmware.run/vmware/vic/5227

Failed it looks like because the container panic'ed and thus was unable to power off.

```
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <6>[    3.366594] tether[207]: segfault at 38 ip 00000000004e8236 sp 000000c42003bab0 error 4 in tether[400000+707000]
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <0>[    3.427570] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000200
2016-09-28T19:09:02.712Z| vcpu-0| I120+ <0>[    3.427570] 
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.429707] CPU: 0 PID: 203 Comm: tether Not tainted 4.4.8-esx #1-photon
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.431193] Hardware name: VMware, Inc. VMware Virtual Platform/440BX Desktop Reference Platform, BIOS 6.00 09/21/2015
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.434055]  0000000000000000 ffffffff81215caf ffffffff816ebe50 ffff880071ebbd18
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.434710]  ffffffff810c7be6 0000000000000010 ffff880071ebbd28 ffff880071ebbcc0
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.435340]  ffff88006f879740 0000000000000200 ffff880079498010 ffff880079497b58
2016-09-28T19:09:02.712Z| vcpu-0| I120: Guest: <4>[    3.435980] Call Trace:
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.436619]  [<ffffffff81215caf>] ? dump_stack+0x5c/0x7d
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.437226]  [<ffffffff810c7be6>] ? panic+0xbf/0x1d7
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.438277]  [<ffffffff81045193>] ? do_exit+0xa63/0xa70
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.439080]  [<ffffffff810451fe>] ? do_group_exit+0x2e/0xa0
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.439664]  [<ffffffff8104dd96>] ? get_signal+0x1b6/0x530
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.440423]  [<ffffffff8100308e>] ? do_signal+0x1e/0x5d0
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.441171]  [<ffffffff81071178>] ? pick_next_task_fair+0x3a8/0x4a0
2016-09-28T19:09:02.713Z| vcpu-0| I120: Guest: <4>[    3.441751]  [<ffffffff8140decf>] ? __schedule+0x38f/0x7e0
2016-09-28T19:09:02.714Z| vcpu-0| I120: Guest: <4>[    3.442835]  [<ffffffff8100104f>] ? exit_to_usermode_loop+0x7f/0xa0
2016-09-28T19:09:02.714Z| vcpu-0| I120: Guest: <4>[    3.443393]  [<ffffffff810013b1>] ? prepare_exit_to_usermode+0x21/0x30
2016-09-28T19:09:02.714Z| vcpu-0| I120: Guest: <4>[    3.443963]  [<ffffffff81411be5>] ? retint_user+0x8/0x13
2016-09-28T19:09:02.714Z| vcpu-0| I120: Guest: <0>[    3.446124] Kernel Offset: disabled
```

[integration_logs_5227_08cb9fd26343bc83f55848b75b586513400f1f8c.zip](https://github.com/vmware/vic/files/499062/integration_logs_5227_08cb9fd26343bc83f55848b75b586513400f1f8c.zip)

@gigawhitlocks I'm not sure where you got that line from - I'm looking in the logs from that test failure and I see the panic from `b679707705480f3fb4ad2ca667c2906ea64c097dfd8f475ae83b75018729a02b`:

```
time="2016-10-07T19:57:10Z" level=debug msg="Launched command with pid 204" 
time="2016-10-07T19:57:10Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/tether.(*tether).launch:338] [231.523042ms] launching session b679707705480f3fb4ad2ca667c2906ea64c097dfd8f475ae83b75018729a02b" 
time="2016-10-07T19:57:10Z" level=info msg="Passing config to Attach" 
time="2016-10-07T19:57:10Z" level=debug msg="[BEGIN] [main.(*attachServerSSH).Reload:71] attach reload" 
time="2016-10-07T19:57:10Z" level=info msg="Processing config for session b679707705480f3fb4ad2ca667c2906ea64c097dfd8f475ae83b75018729a02b" 
time="2016-10-07T19:57:10Z" level=info msg="Session b679707705480f3fb4ad2ca667c2906ea64c097dfd8f475ae83b75018729a02b is configured for attach" 
time="2016-10-07T19:57:10Z" level=debug msg="[BEGIN] [main.(*attachServerSSH).start:116] start attach server" 
time="2016-10-07T19:57:10Z" level=debug msg="[ END ] [main.(*attachServerSSH).start:116] [17.899156ms] start attach server" 
time="2016-10-07T19:57:11Z" level=debug msg="[ END ] [main.(*attachServerSSH).Reload:71] [113.590066ms] attach reload" 
2016/10/07 19:57:11 dispatching power op "OS_Halt"
time="2016-10-07T19:57:10Z" level=debug msg="[0xc42018f940] write \"\\x1b[H\\x1b[JMem: 60452K used, 1992724K free, 11760K shrd, 548K buff, 25668K cached\\nCPU:  6.2% usr 25.0% sys  0.0% nic 62.5% idle  0.0% io  0.0% irq  6.2% sirq\\nLoad average: 0.00 0.00 0.00 2/68 204\\n\\x1b[7m  PID  PPID USER     STAT   VSZ %VSZ CPU %CPU COMMAND\\x1b[0m\\n    1     0 root     S     260m 12.9   1 37.5 /.tether/tether\\n  204     1 root     R     1180  0.0   0  0.0 /bin/top\\n  105     1 root     Z        0  0.0   0  0.0 [systemd-udevd]\\n   83     1 root     Z        0  0.0   0  0.0 [systemd-journal]\\n  151     1 root     Z        0  0.0   0  0.0 [bootstrap]\\n  103     2 root     SW       0  0.0   1  0.0 [kworker/1:4]\\n  176     2 root     SW<      0  0.0   1  0.0 [kworker/1:1H]\\n  171     2 root     SW       0  0.0   0  0.0 [scsi_eh_2]\\n    2     0 root     SW       0  0.0   1  0.0 [kthreadd]\\n    6     2 root     SW       0  0.0   1  0.0 [kworker/u4:0]\\n    5     2 root     SW<      0  0.0   0  0.0 [kworker/0:0H]\\n   10     2 root     SW       0  0.0   1  0.0 [migration/1]\\n   11     2 root     SW       0  0.0   1  0.0 [ksoftirqd/1]\\n    9     2 root     SW       0  0.0   0  0.0 [migration/0]\\n   14     2 root     SW       0  0.0   1  0.0 [kdevtmpfs]\\n   15     2 root     SW<      0  0.0   1  0.0 [netns]\\n   13     2 root     SW<      0  0.0   1  0.0 [kworker/1:0H]\\n   18     2 root     SWN      0  0.0   1  0.0 [ksmd]\\n   19     2 root     SWN      0  0.0   0  0.0 [khugepaged]\\n   20     2 root     SW<      0  0.0   1  0.0 [crypto]\\r\" to 2 writers (err: <nil>)" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x50c0a6]

goroutine 9 [running]:
panic(0x886f40, 0xc42000c050)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/lib/tether.(*Toolbox).halt(0xc4200cf080, 0x20, 0xc42018fe20)
	/drone/src/github.com/vmware/vic/lib/tether/toolbox.go:158 +0x46
github.com/vmware/vic/lib/tether.(*Toolbox).(github.com/vmware/vic/lib/tether.halt)-fm(0xc42000d4f0, 0xc42018fe20)
	/drone/src/github.com/vmware/vic/lib/tether/toolbox.go:102 +0x2a
github.com/vmware/vic/pkg/vsphere/toolbox.(*PowerCommand).Dispatch(0xc4200d32b0, 0x0, 0x0, 0x0, 0x3, 0xc4201bbe40, 0xc4201be191, 0xc42007d450, 0x7)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/power.go:98 +0x30d
github.com/vmware/vic/pkg/vsphere/toolbox.(*PowerCommand).Dispatch-fm(0x0, 0x0, 0x0, 0x7, 0xc42007d4f0, 0xc4201be101, 0x7, 0x7)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/power.go:74 +0x48
github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Dispatch(0xc420056420, 0xc4201be191, 0x7, 0x7, 0x0, 0x0, 0x8)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:191 +0x194
github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Start.func1(0xc420056420)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:135 +0x1f1
created by github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Start
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:145 +0xb1
```

@dougm It looks like we're triggering a power off before toolbox.Reload has been called. That appears to be a permitted path - I cannot confirm without entry/exit trace in toolbox.

It does mean that at that time we've got no idea what the config is so also do not know what the stop signal, or the process PID is. This is another of those scenarios where we should have switched to pre/post calls.

Turns out that recover only works on the goroutine that triggered the panic so we miss that halt path entirely: https://play.golang.org/p/wmIiCLO9r4

@gigawhitlocks where did your trace come from? Looking at it it seems likely that the line numbers are out by 1 given we call Stack on line 37, not line 36. 
The panic you comment with is actually printed in the recover block, not from the raw go panic handler. This would have been recovered from because we then call halt.

We only get the output of one thread because default behaviour changed in 1.6: http://dave.cheney.net/tag/gotraceback

Why you're panic'ing I cannot say - only thing I can think of is that the Open failed so logFile is nil

Seen again in https://ci.vmware.run/vmware/vic/5543/1
VCH: VCH-5543-3468
Container: hTp1aXImLo2Gpu3

```
time="2016-10-08T10:13:33Z" level=debug msg="Resolved /bin/top to /bin/top" 
time="2016-10-08T10:13:33Z" level=debug msg="[0xc420173240] Going into sleep with 0 readers" 
time="2016-10-08T10:13:33Z" level=debug msg="Launched command with pid 200" 
time="2016-10-08T10:13:33Z" level=debug msg="[ END ] [github.com/vmware/vic/lib/tether.(*tether).launch:338] [166.617037ms] launching session fa33431b60a9ae25e1a23c6398eb215d3f28e34525e66162c606b4bbd1493d18" 
time="2016-10-08T10:13:33Z" level=info msg="Passing config to Attach" 
time="2016-10-08T10:13:33Z" level=debug msg="[BEGIN] [main.(*attachServerSSH).Reload:71] attach reload" 
time="2016-10-08T10:13:33Z" level=info msg="Processing config for session fa33431b60a9ae25e1a23c6398eb215d3f28e34525e66162c606b4bbd1493d18" 
time="2016-10-08T10:13:33Z" level=info msg="Session fa33431b60a9ae25e1a23c6398eb215d3f28e34525e66162c606b4bbd1493d18 is configured for attach" 
time="2016-10-08T10:13:34Z" level=debug msg="[BEGIN] [main.(*attachServerSSH).start:116] start attach server" 
time="2016-10-08T10:13:33Z" level=debug msg="[0xc4201792a0] write \"\\x1b[H\\x1b[JMem: 60708K used, 1992468K free, 11760K shrd, 548K buff, 25668K cached\\nCPU: 13.3% usr 20.0% sys  0.0% nic 66.6% idle  0.0% io  0.0% irq  0.0% sirq\\nLoad average: 0.00 0.00 0.00 2/66 202\\n\\x1b[7m  PID  PPID USER     STAT   VSZ %VSZ CPU %CPU COMMAND\\x1b[0m\\n    1     0 root     S     394m 19.6   0 33.3 /.tether/tether\\n  200     1 root     R     1180  0.0   1  0.0 /bin/top\\n   77     1 root     Z        0  0.0   0  0.0 [systemd-journal]\\n  109     1 root     Z        0  0.0   1  0.0 [systemd-udevd]\\n  148     1 root     Z        0  0.0   1  0.0 [bootstrap]\\n   12     2 root     SW       0  0.0   1  0.0 [kworker/1:0]\\n  167     2 root     SW       0  0.0   0  0.0 [scsi_eh_2]\\n  182     2 root     SW<      0  0.0   1  0.0 [kworker/1:1H]\\n    9     2 root     SW       0  0.0   0  0.0 [migration/0]\\n    2     0 root     SW       0  0.0   1  0.0 [kthreadd]\\n    7     2 root     SW       0  0.0   1  0.0 [rcu_sched]\\n    4     2 root     SW       0  0.0   0  0.0 [kworker/0:0]\\n   13     2 root     SW<      0  0.0   1  0.0 [kworker/1:0H]\\n    6     2 root     SW       0  0.0   0  0.0 [kworker/u4:0]\\n   15     2 root     SW<      0  0.0   1  0.0 [netns]\\n    8     2 root     SW       0  0.0   0  0.0 [rcu_bh]\\n   17     2 root     SW<      0  0.0   1  0.0 [writeback]\\n   14     2 root     SW       0  0.0   1  0.0 [kdevtmpfs]\\n   19     2 root     SWN      0  0.0   1  0.0 [khugepaged]\\n   20     2 root     SW<      0  0.0   0  0.0 [crypto]\\r\" to 2 writers (err: <nil>)" 
2016/10/08 10:13:34 dispatching power op "OS_Halt"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x50c1d6]

goroutine 10 [running]:
panic(0x8870e0, 0xc42000c050)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/vmware/vic/lib/tether.(*Toolbox).halt(0xc4200d1080, 0x20, 0xc420130b20)
	/drone/src/github.com/vmware/vic/lib/tether/toolbox.go:158 +0x46
github.com/vmware/vic/lib/tether.(*Toolbox).(github.com/vmware/vic/lib/tether.halt)-fm(0xc42000d4f0, 0xc420130b20)
	/drone/src/github.com/vmware/vic/lib/tether/toolbox.go:102 +0x2a
github.com/vmware/vic/pkg/vsphere/toolbox.(*PowerCommand).Dispatch(0xc42008b2b0, 0x0, 0x0, 0x0, 0x3, 0xc420130aa0, 0xc4201379d1, 0xc42007d380, 0x7)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/power.go:98 +0x30d
github.com/vmware/vic/pkg/vsphere/toolbox.(*PowerCommand).Dispatch-fm(0x0, 0x0, 0x0, 0x7, 0xc42007d420, 0xc420137901, 0x7, 0x7)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/power.go:74 +0x48
github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Dispatch(0xc420056420, 0xc4201379d1, 0x7, 0x7, 0x0, 0x0, 0x8)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:191 +0x194
github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Start.func1(0xc420056420)
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:135 +0x1f1
created by github.com/vmware/vic/pkg/vsphere/toolbox.(*Service).Start
	/drone/src/github.com/vmware/vic/pkg/vsphere/toolbox/service.go:145 +0xb1
```

--- a/lib/tether/toolbox.go
+++ b/lib/tether/toolbox.go
@@ -20,6 +20,7 @@ import (
 	"errors"
 	"fmt"
 	"net"
+	"sync"
 	"syscall"
 	"time"
 
@@ -35,8 +36,12 @@ import (
 type Toolbox struct {
 	*toolbox.Service
 
-	config *ExecutorConfig
-	stop   chan struct{}
+	sess struct {
+		sync.Mutex
+		session *SessionConfig
+	}
+
+	stop chan struct{}
 }
 
 // NewToolbox returns a tether.Extension that wraps the vsphere/toolbox service
@@ -93,7 +98,12 @@ func (t *Toolbox) Stop() error {
 
 // Reload implementation of the tether.Extension interface
 func (t *Toolbox) Reload(config *ExecutorConfig) error {
-	t.config = config
+	if config != nil && config.Sessions != nil {
+		t.sess.Lock()
+		defer t.sess.Unlock()
+		t.sess.session = config.Sessions[config.ID]
+	}
+
 	return nil
 }
 
@@ -108,9 +118,24 @@ func (t *Toolbox) InContainer() *Toolbox {
 	return t
 }
 
-func (t *Toolbox) kill(name string) error {
-	session := t.config.Sessions[t.config.ID]
+func (t *Toolbox) session() *SessionConfig {
+	t.sess.Lock()
+	defer t.sess.Unlock()
+	return t.sess.session
+}
 
+func (t *Toolbox) kill(name string) error {
+	session := t.session()
+	if session == nil {
+		return fmt.Errorf("failed to kill container: process not found")
+	}
+
+	session.m.Lock()
+	defer session.m.Unlock()
+	return t.killHelper(session, name)
+}
+
+func (t *Toolbox) killHelper(session *SessionConfig, name string) error {
 	if name == "" {
 		name = string(ssh.SIGTERM)
 	}
@@ -137,8 +162,17 @@ func (t *Toolbox) containerAuthenticate(_ toolbox.VixCommandRequestHeader, data 
 	if err := c.UnmarshalBinary(data); err != nil {
 		return err
 	}
+
+	session := t.session()
+	if session == nil {
+		return errors.New("not yet initialized")
+	}
+
+	session.m.Lock()
+	defer session.m.Unlock()
+
 	// no authentication yet, just using container ID as a sanity check for now
-	if c.Name != t.config.ID {
+	if c.Name != session.ID {
 		return errors.New("failed to verify container ID")
 	}
 
@@ -155,10 +189,17 @@ func (t *Toolbox) containerStartCommand(r *toolbox.VixMsgStartProgramRequest) (i
 }
 
 func (t *Toolbox) halt() error {
-	session := t.config.Sessions[t.config.ID]
+	session := t.session()
+	if session == nil {
+		return fmt.Errorf("failed to halt container: not initialized yet")
+	}
+
+	session.m.Lock()
+	defer session.m.Unlock()
+
 	log.Infof("stopping %s", session.ID)
 
-	if err := t.kill(session.StopSignal); err != nil {
+	if err := t.killHelper(session, session.StopSignal); err != nil {
 		return err
 	}
 

